lib/spack/spack/test/stage.py

```diff
@@ -728,21 +741,96 @@ def test_first_accessible_path(self, tmpdir):
         # Cleanup
         shutil.rmtree(str(name))
 
+    def test_create_stage_root(self, tmpdir, no_path_access):
+        """Test _create_stage_root permissions."""
+        test_dir = tmpdir.join('path')
+        test_path = str(test_dir)
+
+        try:
+            if getpass.getuser() in str(test_path).split(os.sep):
+                # Simply ensure directory created if tmpdir includes user
+                spack.stage._create_stage_root(test_path)
+                assert os.path.exists(test_path)
+
+                p_stat = os.stat(test_path)
+                assert p_stat.st_mode & stat.S_IRWXU == stat.S_IRWXU
+            else:
+                # Ensure an OS Error is raised on created, non-user directory
+                with pytest.raises(OSError) as exc_info:
+                    spack.stage._create_stage_root(test_path)
+
+                assert exc_info.value.errno == errno.EACCES
+        finally:
+            try:
+                shutil.rmtree(test_path)
+            except OSError:
+                pass
+
+    @pytest.mark.nomockstage
+    def test_create_stage_root_bad_uid(self, tmpdir, monkeypatch):
+        """
+        Test the code path that uses an existing user path -- whether `$user`
+        in `$tempdir` or not -- and triggers the generation of the UID
+        mismatch warning.
+
+        This situation can happen with some `config:build_stage` settings
+        for teams using a common service account for installing software.
+        """
+        orig_stat = os.stat
+
+        class MinStat:
+            st_mode = -1
+            st_uid = -1
+
+        def _stat(path):
+            p_stat = orig_stat(path)
+
+            fake_stat = MinStat()
+            fake_stat.st_mode = p_stat.st_mode
+            return fake_stat
+
+        user_dir = tmpdir.join(getpass.getuser())
+        user_dir.ensure(dir=True)
+        user_path = str(user_dir)
+
+        # TODO: If we could guarantee access to the monkeypatch context
+        # function (i.e., 3.6.0 on), the call and assertion could be moved
+        # to a with block, such as:
+        #
+        #  with monkeypatch.context() as m:
```

Comment:
Overall I'm not sure what the difference between these methods (i.e. why would this approach be more desirable)

Reply:
True that it is still patching `os` calls, but it does provide finer grained control of the patch by limiting its scope to the `with` code block.

Reply:
Given that 3.5 will be deprecated as of the end of next year (https://devguide.python.org/#status-of-python-branches) I suppose this could stick around on that count. But I think the better thing to do is to not touch `os.stat` at all (as in https://github.com/spack/spack/pull/12733#discussion_r327748097). So if there is a TODO here I'd rather it said "don't monkeypatch os.stat".

Reply:
I understand but monkeypatching `os.stat` is debatable.  Todd (tgamblin) liked the idea (October 8th meeting) given the difficulty of reproducing paths with different GIDs, etc.